### PR TITLE
dissallow llvm-openmp 12.* for M1

### PR DESCRIPTION
**Pull request by esc**
The `llvm-openmp` package has a Numba relevant bug that only manifests
when using that package on M1 silicon (new Apple chips) in version 12.
This will allow building Numba (the headers are needed) but dissallow
running it with `llvm-openmp` in version 12.* on Apple M1 silicon.

The `llvm-openmp` bug is here:

https://github.com/llvm/llvm-project/issues/49923

This will also be future proof and allow `llvm-openmp` to be installed
in version 13 as soon as that becomes available from `main/defaults` or
if people want to use `conda-forge`, since that already has version 13.

### CODE REVIEW

File: buildscripts/condarecipe.local/meta.yaml [hunk 1]
```diff
@@ -70,7 +71,7 @@ test:
     - ipython                  # [not (armv6l or armv7l or aarch64)]
     - setuptools
     - tbb  >=2021              # [not (armv6l or armv7l or aarch64 or linux32 or ppc64le)]
-    - llvm-openmp              # [osx]
+    - llvm-openmp              # [osx and x86_64]
```
Comment (stuartarchibald) [inline buildscripts/condarecipe.local/meta.yaml:74]: This is banning all testing with llvm-openmp on M1? Won't the `run_constrained` just handle this?

Comment (stuartarchibald) [reply buildscripts/condarecipe.local/meta.yaml:74]: I guess this comes down to the definition of the "test environment". IIRC in principle the `meta.yaml` should be a description of what is correct and it's up to e.g. `conda build --test` to have the appropriate channels etc to pick up the packages. In scenarios where testing is performed from the installed package directly, it's a question of ensuring that environment is suitable if OpenMP testing is required.

**Comment by esc** [comment]
Just focusing on the Python 3.10 and Numpy 1.21 combo to at least get something going.

I can build this with:

```
conda build -c numba/label/ci --dirty --no-test --python=3.10 --numpy=1.21 buildscripts/condarecipe.local
```

We need to get the `llvmlite` 0.39 dev package from `numba/label/ci` b/c the M1 package never made it to `numba/label/dev`.

**Comment by esc** [comment]
I have now reduced the patch to basically:

```
  run_constrained:
    ...
    - llvm-openmp > 12.0.0      # [arm64]
```

I.e. just the change to the `run_constrained`. Yet still, when I go to try to resolve and install the resulting package, I get:

```
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: /
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.                                                                                                               failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package llvmlite conflicts for:
numba==0.56.0dev0=np1.21py3.10h5ae3daa_gd4967d87c_254 -> llvmlite[version='>=0.39.0dev0,<0.39']
numba/label/ci|numba/label/ci::llvmlite
numba/label/ci::llvmlite

Package numpy conflicts for:
numba==0.56.0dev0=np1.21py3.10h5ae3daa_gd4967d87c_254 -> numpy[version='>=1.18']
numpy=1.21

Package setuptools conflicts for:
numba==0.56.0dev0=np1.21py3.10h5ae3daa_gd4967d87c_254 -> setuptools
python=3.10 -> pip -> setuptools

Package python conflicts for:
numba/label/ci::llvmlite -> python[version='>=3.10,<3.11.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0']
numba==0.56.0dev0=np1.21py3.10h5ae3daa_gd4967d87c_254 -> llvmlite[version='>=0.39.0dev0,<0.39'] -> python[version='>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
numpy=1.21 -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.10
numba==0.56.0dev0=np1.21py3.10h5ae3daa_gd4967d87c_254 -> python[version='>=3.10,<3.11.0a0']

Package expat conflicts for:
numpy=1.21 -> python[version='>=3.9,<3.10.0a0'] -> expat[version='>=2.4.1,<3.0a0']
numba/label/ci::llvmlite -> python[version='>=3.9,<3.10.0a0'] -> expat[version='>=2.4.1,<3.0a0']
```

I have also manually verified that this doesn't happen if the `run_constrained` entry about `open-llvm` is omitted. I.e. the package will seemingly build fine then.

I am now wondering if there are any additional changes to the `meta.yaml` that need to be made.

**Comment by esc** [comment]
Taking a closer look at the "rendered" `meta.yaml` -- which I obtained from unpacking the conda package and looking in `info` has the following stanzas:

```
  run:
    - libcxx >=12.0.0
    - llvm-openmp >=12.0.0
    - llvmlite >=0.39.0dev0,<0.39
    - numpy >=1.18
    - python >=3.10,<3.11.0a0
    - setuptools
  run_constrained:
    - cuda-python >=11.6
    - cudatoolkit >=9.2
    - llvm-openmp > 12.0.0
    - scipy >=1.0
    - tbb >=2021
```

I see `llvm-openmp` in both `run` and `run_constrained`, yet with potentially conflicting version specifiers. This may explain the strange conflict explanations from `conda` that don't hint at anything.

Current assumption is that `llvm-openmp` is "bleeding through" from the `build` section to the `run` section.

**Comment by esc** [comment]
I investigated this issue some more with @seibert last night and I managed to work out (I think) why the `llvm-openmp` requirement was bleeding through to the `run` section. So, the recipe of `llvm-openmp`, at least the one on conda-forge, has something called a `run_export`(https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements). See also:  https://github.com/conda-forge/openmp-feedstock/blob/77d3f5b/recipe/meta.yaml#L28-L32 . Effectively this means that if `llvm-openmp` is included somewhere in `build`, then it is automatically exported as a `run` requirement and this behaviour is inherited into anything requiring `llvm-openmp` in `build`.And, yes, we can use `conda render` to check this.

OK, so far, so good. Now, there is a way to disable this, using the `ignore_run_exports` section in `build` (not `requirements`).

So for the Numba recipe this would be:

```
build:
  ignore_run_exports_from:
    - llvm-openmp
```

In fact, this technique has been used before for `tbb` in the Numba recipe! So we can use the following patch to disable exporting `llvm-openmp`:

```
diff --git a/buildscripts/condarecipe.local/meta.yaml b/buildscripts/condarecipe.local/meta.yaml
index bc1c505ae..4c5ace351 100644
--- a/buildscripts/condarecipe.local/meta.yaml
+++ b/buildscripts/condarecipe.local/meta.yaml
@@ -20,6 +20,7 @@ build:
   ignore_run_exports:
     # tbb-devel triggers hard dependency on tbb, this is not the case.
     - tbb     # [not (armv6l or armv7l or aarch64 or linux32)]
+    - llvm-openmp

 requirements:
   # build and run dependencies are duplicated to avoid setuptools issues
@@ -29,6 +30,7 @@ requirements:
     - {{ compiler('cxx') }}    # [not (armv6l or armv7l or aarch64)]
     # OpenMP headers from llvm needed for OSX.
     - llvm-openmp              # [osx]
+
   host:
     - python
     - numpy
@@ -70,7 +72,7 @@ test:
     - ipython                  # [not (armv6l or armv7l or aarch64)]
     - setuptools
     - tbb  >=2021              # [not (armv6l or armv7l or aarch64 or linux32 or ppc64le)]
-    - llvm-openmp              # [osx]
+    - llvm-openmp              # [osx and not arm64]
     # This is for driving gdb tests
     - pexpect                  # [linux64]
     # For testing ipython
```

Using `conda render` we get the following:

```
  run:
    - setuptools
    - numpy >=1.18
    - llvmlite >=0.39.0dev0,<0.39
    - libcxx >=12.0.0
    - python >=3.10,<3.11.0a0
  run_constrained:
    - scipy >=1.0
    - cuda-python >=11.6
    - tbb >=2021
    - cudatoolkit >=9.2
```

And this is all fine and it builds a package and so on. However, when we then go and attempt to install it, `llvm-openmp` is still pulled in as a dependency. When I try to remove it from the environment, I get:

```
The following packages will be REMOVED:

  blas-1.0-openblas
  libgfortran-5.0.0-11_1_0_h6a59814_26
  libgfortran5-11.1.0-h6a59814_26
  libopenblas-0.3.17-hea475bc_1
  llvm-openmp-12.0.0-haf9daa7_1
  numba-0.56.0dev0-np1.21py3.10h5ae3daa_gaed25b8aa_257
  numpy-1.21.2-py310hb38b75b_0
  numpy-base-1.21.2-py310h6269429_0
```

And we can remove `numba` but the other packages, `blas, libgfortran, libgfortran5, libopenblas, llvm-openmp, numpy numpy-base` are all-or-nothing. So, from that my conclusion is that we can not install Numba as a conda package w/o also installing `llvm-openmp`.

For what it is worth we can attempt to add `llvm-openmp` to `run_constrained`:

```
diff --git a/buildscripts/condarecipe.local/meta.yaml b/buildscripts/condarecipe.local/meta.yaml
index bc1c505ae..3357e7756 100644
--- a/buildscripts/condarecipe.local/meta.yaml
+++ b/buildscripts/condarecipe.local/meta.yaml
@@ -20,6 +20,7 @@ build:
   ignore_run_exports:
     # tbb-devel triggers hard dependency on tbb, this is not the case.
     - tbb     # [not (armv6l or armv7l or aarch64 or linux32)]
+    - llvm-openmp

 requirements:
   # build and run dependencies are duplicated to avoid setuptools issues
@@ -29,6 +30,7 @@ requirements:
     - {{ compiler('cxx') }}    # [not (armv6l or armv7l or aarch64)]
     # OpenMP headers from llvm needed for OSX.
     - llvm-openmp              # [osx]
+
   host:
     - python
     - numpy
@@ -51,6 +53,7 @@ requirements:
   run_constrained:
     # If TBB is present it must be at least version 2021
     - tbb >=2021    # [not (armv6l or armv7l or aarch64 or linux32 or ppc64le)]
+    - llvm-openmp >=13.0.0    # [arm64]
     # avoid confusion from openblas bugs
     - libopenblas !=0.3.6      # [x86_64]
     # CUDA 9.2 or later is required for CUDA support
@@ -70,7 +73,7 @@ test:
     - ipython                  # [not (armv6l or armv7l or aarch64)]
     - setuptools
     - tbb  >=2021              # [not (armv6l or armv7l or aarch64 or linux32 or ppc64le)]
-    - llvm-openmp              # [osx]
+    - llvm-openmp              # [osx and not arm64]
     # This is for driving gdb tests
     - pexpect                  # [linux64]
     # For testing ipython
```

But then we run into the following conflict (which incidentally says nothing about `llvm-openmp`:

```
UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package setuptools conflicts for:
python=3.10 -> pip -> setuptools
numba==0.56.0dev0=np1.21py3.10h5ae3daa_ge269044c9_260 -> setuptools

Package libiconv conflicts for:
numpy=1.21 -> python[version='>=3.9,<3.10.0a0'] -> libiconv[version='>=1.16,<2.0a0']
numba/label/ci::llvmlite -> python[version='>=3.9,<3.10.0a0'] -> libiconv[version='>=1.16,<2.0a0']

Package numpy conflicts for:
numba==0.56.0dev0=np1.21py3.10h5ae3daa_ge269044c9_260 -> numpy[version='>=1.18']
numpy=1.21

Package python conflicts for:
numpy=1.21 -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
numba==0.56.0dev0=np1.21py3.10h5ae3daa_ge269044c9_260 -> python[version='>=3.10,<3.11.0a0']
numba/label/ci::llvmlite -> python[version='>=3.10,<3.11.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0']
numba==0.56.0dev0=np1.21py3.10h5ae3daa_ge269044c9_260 -> llvmlite[version='>=0.39.0dev0,<0.39'] -> python[version='>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.10

Package llvmlite conflicts for:
numba==0.56.0dev0=np1.21py3.10h5ae3daa_ge269044c9_260 -> llvmlite[version='>=0.39.0dev0,<0.39']
numba/label/ci|numba/label/ci::llvmlite
numba/label/ci::llvmlite
```


Finally, my current take on this is that we can not disallow using a broken `llvm-openmp` 12.0.0 to be installed alongside Numba with the current state of `defaults`. Having 13.0.0 on `defaults` is likely to solve this issue w/o any need to alter or change Numba or how it is built.

**Comment by esc** [comment]
@stuartarchibald I think it is safe to say, that we can abandon this one?

**Pull request closed without merging**